### PR TITLE
add per-talk type submission deadline and option for late submissions

### PR DESCRIPTION
**Pull request by terceiro**
Sometime conferences want to continue to receive submissions after the
deadline, and maybe accept them if there is space and they are
interesting enough. This patch displays them as being late submissions
in both the talk list and in the talk page, for the benefit of
reviewers and content admins.

This feature is completely opt-in: if WAFER_TALKS_DEADLINE is not set,
nothing changes.

### CODE REVIEW

**Comment by terceiro** [comment]
I have seen that the tests fail on python2 because I'm using unittest.mock. I can easily amend my patch to fallback to the top-level mock module, which is even already in requirements-dev.txt. but there is a larger point which is: should we really still care about python2 at this point?

**Comment by terceiro** [comment]
> should we really still care about python2 at this point?

you know what, nevermind. my point still stands, but I don't want this PR to be about anything other than the actual feature I am proposing.

**Comment by drnlm** [comment]
In general, I like the idea.

However, I am concerned that a single deadline is too simplistic. We have multiple talk types, and may want different deadlines for each of them. For PyCon ZA, for example, we usually close tutorial submissions well before closing talk submissions (which is why we have support for individually disabling sugmissions of each talk type).

**Comment by terceiro** [comment]
yeah, having this be per-track would be great. can you point me to the code that implements that? I skimmed the pyconza2019 repo but couldn't find it.

**Comment by hodgestar** [comment]
One option might be to put the deadline into the TalkType model (if we feel that talk type is the correct way to break up the different deadlines) -- https://github.com/CTPUG/wafer/blob/master/wafer/talks/models.py#L51-L77

**Comment by terceiro** [comment]
argh I confused talk type with track. yes IMO it makes more sense to do this by talk type (and in there indeed there is separate submission flag)

**Comment by terceiro** [comment]
ok, so I reworked my original idea, and now added a submission_deadline field and a "accept late submissions" flag to TalkType. I then replaced all direct usage of disable_submission with a new method in TalkType.objects that fetches all talk types that can have talks submitted to them.

File: wafer/talks/models.py [hunk 1]
```diff
@@ -47,6 +48,21 @@ def authors_help():
     ))
     return ' '.join(text)
 
+class TalkTypeManager(models.Manager):
+
+    def open_for_submission(self, plus=models.Q(pk__lt=0)):
```
Comment (hodgestar) [inline wafer/talks/models.py]: I would remove the `plus` parameter and simply allow the caller to do their own `| Q(pk=initial) `.

File: wafer/talks/tests/test_talk_type.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+def test_open_by_default():
+    from wafer.talks.models import TalkType
```
Comment (hodgestar) [inline wafer/talks/tests/test_talk_type.py]: I assume this strange importing of things inside the test functions was copied from `test_wafer_basic_talks.py`? If so, does anyone remember why it is like that? If not, perhaps we should just try importing things the ordinary way at the top of the module?

File: wafer/talks/templates/wafer.talks/talk.html [hunk 3]
```diff
@@ -108,6 +108,11 @@ <h1>
         {% else %}
           <span class="badge badge-danger">{% trans 'Not accepted' %}</span>
         {% endif %}
+        {% if object.is_late_submission %}
+        {% with submission=object.submission_time %}
+        <span class="badge badge-warning ">{% blocktrans %}Late submission ({{ submission }}){% endblocktrans %}</span>
```
Comment (hodgestar) [inline wafer/talks/templates/wafer.talks/talk.html]: Should we display the submission time somewhere even if the talk was not submitted late?

File: wafer/talks/tests/test_talk_type.py (hunk 2 above)
Comment (drnlm) [reply wafer/talks/tests/test_talk_type.py]: This pattern is from the first tests we wrote  - I think we needed to work around some odd behaviour with how importing `django.contrib.auth` worked when running the tests at the time, and just never updated it once that was no longer needed.

File: wafer/talks/models.py (hunk 1 above)
Comment (terceiro) [reply wafer/talks/models.py]: done

File: wafer/talks/tests/test_talk_type.py (hunk 2 above)
Comment (terceiro) [reply wafer/talks/tests/test_talk_type.py]: ack, I moved the imports to the top

File: wafer/talks/templates/wafer.talks/talk.html (hunk 3 above)
Comment (terceiro) [reply wafer/talks/templates/wafer.talks/talk.html]: good idea

**Comment by terceiro** [comment]
hello, I just addressed all of the comments, and pushed a new version of this PR

**Comment by terceiro** [comment]
nice, thanks! it would be cool to publish a new release, there's been quite some work in master since the last one

**Pull request merged**